### PR TITLE
chore(istio): add unit tests for ServiceEntries

### PR DESCRIPTION
**Pull request by jewertow**

### CODE REVIEW

**Review by bartoszmajsak — COMMENTED** [review]
Thanks for extending the test coverage and improving readability. I have just two minor things for you to consider.

File: internal/pkg/istio/config_factory_test.go [hunk 1]
```diff
@@ -241,48 +283,99 @@ func TestEnvoyFilters(t *testing.T) {
...
+		importedServices:          []*v1alpha1.ExportedService{importedSvcAns1, importedSvcBns1, importedSvcAns2},
+		expectedServiceEntryFiles: []string{"dns/fds.yaml", "dns/svc-b-ns-1.yaml", "dns/svc-a-ns-2.yaml"},
+	}}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			client := fake.NewSimpleClientset()
+			informerFactory := informers.NewSharedInformerFactory(client, 0)
+			serviceInformer := informerFactory.Core().V1().Services().Informer()
+			serviceLister := informerFactory.Core().V1().Services().Lister()
+			stopCh := make(chan struct{})
```
[the remark this answers is not in the document]
Comment (jewertow) [reply internal/pkg/istio/config_factory_test.go:340]: Why would we need to care about that in tests? I applied the suggested change, but I got timeouts...

[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply internal/pkg/istio/config_factory_test.go:340]: Interesting, as I did ran it locally before suggesting it and it worked. 

<details>
<summary>$ make test</summary>

```sh
❯ git --no-pager diff --patch-with-raw          
:100644 100644 ca9fd32 0000000 M	internal/pkg/istio/config_factory_test.go

diff --git a/internal/pkg/istio/config_factory_test.go b/internal/pkg/istio/config_factory_test.go
index ca9fd32..9b3df3b 100644
--- a/internal/pkg/istio/config_factory_test.go
+++ b/internal/pkg/istio/config_factory_test.go
@@ -212,6 +212,9 @@ func TestIngressGateway(t *testing.T) {
 			serviceInformer := informerFactory.Core().V1().Services().Informer()
 			serviceLister := informerFactory.Core().V1().Services().Lister()
 			stopCh := make(chan struct{})
+			defer func() {
+				close(stopCh)
+			}()
 			informerFactory.Start(stopCh)
 
 			for _, svc := range tc.localServices {
@@ -267,6 +270,9 @@ func TestEnvoyFilters(t *testing.T) {
 			serviceInformer := informerFactory.Core().V1().Services().Informer()
 			serviceLister := informerFactory.Core().V1().Services().Lister()
 			stopCh := make(chan struct{})
+			defer func() {
+				close(stopCh)
+			}()
 			informerFactory.Start(stopCh)
 
 			for _, svc := range tc.localServices {
@@ -338,6 +344,9 @@ func TestServiceEntries(t *testing.T) {
 			serviceInformer := informerFactory.Core().V1().Services().Informer()
 			serviceLister := informerFactory.Core().V1().Services().Lister()
 			stopCh := make(chan struct{})
+			defer func() {
+				close(stopCh)
+			}()
 			informerFactory.Start(stopCh)
 
 			for _, svc := range tc.localServices {

federation on  unit-tests-for-config-factory [$!] via  v1.23.4 
❯ make                                
go get /home/bartek/code/redhat/ossm/federation/...
go build -C /home/bartek/code/redhat/ossm/federation/cmd/federation-controller -o /home/bartek/code/redhat/ossm/federation/out/federation-controller 
Adding license to /home/bartek/code/redhat/ossm/federation/internal/api/federation/v1alpha1/exported_service.pb.go
Adding license to /home/bartek/code/redhat/ossm/federation/internal/api/federation/v1alpha1/exported_service_deepcopy.gen.go
/home/bartek/code/redhat/ossm/federation/bin/goimports -local "github.com/openshift-service-mesh/federation" -w /home/bartek/code/redhat/ossm/federation/
go test /home/bartek/code/redhat/ossm/federation/...
?   	github.com/openshift-service-mesh/federation/cmd/federation-controller	[no test files]
?   	github.com/openshift-service-mesh/federation/internal/api/federation/v1alpha1	[no test files]
?   	github.com/openshift-service-mesh/federation/internal/pkg/common	[no test files]
?   	github.com/openshift-service-mesh/federation/internal/pkg/config	[no test files]
ok  	github.com/openshift-service-mesh/federation/internal/pkg/fds	(cached)
?   	github.com/openshift-service-mesh/federation/internal/pkg/kube	[no test files]
?   	github.com/openshift-service-mesh/federation/internal/pkg/networking	[no test files]
?   	github.com/openshift-service-mesh/federation/internal/pkg/openshift	[no test files]
?   	github.com/openshift-service-mesh/federation/internal/pkg/xds	[no test files]
?   	github.com/openshift-service-mesh/federation/internal/pkg/xds/adsc	[no test files]
?   	github.com/openshift-service-mesh/federation/internal/pkg/xds/adss	[no test files]
ok  	github.com/openshift-service-mesh/federation/internal/pkg/informer	(cached)
ok  	github.com/openshift-service-mesh/federation/internal/pkg/istio	(cached)

```

</details>

The reason being - test is code so my line of thinking is that we treat it the same as far as correctness goes.

[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply internal/pkg/istio/config_factory_test.go:340]: But there might be something flaky/race condition there.  I ran it 20 times in a row and once I got this:

```sh
2024-12-19T09:54:37.269534Z	fatal	controller	timed out waiting for Service controller caches to sync
```

Ignore my suggestion there. Perhaps we could improve it when controller-runtime comes to the picture with passing context.Context from the tests instead.

**Review by bartoszmajsak — APPROVED** [review]
LGTM

**Pull request merged**